src/IntrusivePtr.h

```diff
@@ -0,0 +1,216 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+#pragma once
+
+#include <type_traits>
+#include <utility>
+
+// An intrusive, reference counting smart pointer implementation.
+template <class T>
+class IntrusivePtr {
+public:
+	// -- member types
+
+	using pointer = T*;
+
+	using const_pointer = const T*;
+
+	using element_type = T;
+
+	using reference = T&;
+
+	using const_reference = const T&;
+
+	// -- constructors, destructors, and assignment operators
+
+	constexpr IntrusivePtr() noexcept : ptr_(nullptr)
+		{
+		// nop
+		}
+
+	constexpr IntrusivePtr(std::nullptr_t) noexcept : IntrusivePtr()
+		{
+		// nop
+		}
+
+	IntrusivePtr(pointer raw_ptr, bool add_ref) noexcept
+		{
+		setPtr(raw_ptr, add_ref);
+		}
+
+	IntrusivePtr(IntrusivePtr&& other) noexcept : ptr_(other.detach())
+		{
+		// nop
+		}
+
+	IntrusivePtr(const IntrusivePtr& other) noexcept
+		{
+		setPtr(other.get(), true);
+		}
+
+	template <class U, class = std::enable_if_t<std::is_convertible_v<U*, T*>>>
+	IntrusivePtr(IntrusivePtr<U> other) noexcept : ptr_(other.detach())
+		{
+		// nop
+		}
+
+	~IntrusivePtr()
+		{
+		if ( ptr_ )
+			Unref(ptr_);
+		}
+
+	void swap(IntrusivePtr& other) noexcept
+		{
+		std::swap(ptr_, other.ptr_);
+		}
+
+	// Returns the raw pointer without modifying the reference count and sets
+	// this to `nullptr`.
+	pointer detach() noexcept
+		{
+		auto result = ptr_;
+		if ( result )
+			ptr_ = nullptr;
+		return result;
+		}
+
+	void reset(pointer new_value = nullptr, bool add_ref = true) noexcept
+		{
+		auto old = ptr_;
+		setPtr(new_value, add_ref);
+		if ( old )
+			Unref(old);
+		}
+
+	IntrusivePtr& operator=(IntrusivePtr other) noexcept
+		{
+		swap(other);
+		return *this;
+		}
+
+	pointer get() const noexcept
+		{
+		return ptr_;
+		}
+
+	pointer operator->() const noexcept
+		{
+		return ptr_;
+		}
+
+	reference operator*() const noexcept
+		{
+		return *ptr_;
+		}
+
+	bool operator!() const noexcept
+		{
+		return !ptr_;
+		}
+
+	explicit operator bool() const noexcept
+		{
+		return ptr_ != nullptr;
+		}
+
+private:
+	void setPtr(pointer raw_ptr, bool add_ref) noexcept
+		{
+		ptr_ = raw_ptr;
+		if ( raw_ptr && add_ref )
+			Ref(raw_ptr);
+		}
+
+	pointer ptr_;
+};
+
+// Convenience function for creating intrusive pointers.
+template <class T, class... Ts>
+IntrusivePtr<T> makeCounted(Ts&&... args)
```

Comment:
`makeIntrusive` seems more natural? Or actually even `make_intrusive` for consistency with standard smart pointers.

Reply:
I don't have a strong preference on the name. What I like about `makeCounted` is that focuses on the *object*: "make a (reference) counted object". Much like the STL functions: `make_shared` for "make a shared object" and `make_unique` for "make a uniquely owned object".

Since we use CamelCase for the class, I think it's more consistent if the factory function also follows that. But there are also many functions in Zeek that use std_case.

So... whichever you guys prefer / is more aligned with the Zeek style. 😉

Reply:
The style is mostly that global functions are snake case and class methods are CamelCase, from what I can tell. I didn't spell that out in the style guide though and it should probably be fixed.